aklock.man
.TH AKLOCK 1
.SH NAME
aklock \- Lock X display until password supplied
.SH SYNOPSIS
.B aklock
.SH DESCRIPTION
.B aklock
locks the X server till the user enters their password at the keyboard.

While
.B aklock
is running, the mouse and keyboard are grabbed and the mouse cursor
becomes a padlock.  Output displayed by X programs, and windows put up
by new X clients, continue to be visible, and any new output is
displayed normally, except you run in -blank mode. Then the content
of the Screen is hidden behind a black Window.

The mouse and keyboard are returned when the user types their
password, followed by Enter or Newline.  If an incorrect password is
entered the bell is sounded.  Pressing Backspace or Delete erases one
character of a password partially typed; pressing Escape or Clear
clears anything that has been entered.

If too many attempts are made in too short a time further keystrokes
generate bells and are otherwise ignored until a timeout has expired.

The X server screen saver continues to operate normally; if it comes
into operation the display may be restored by the usual means of
touching a key (Shift, for example) or the mouse.
.SH OPTIONS, X RESOURCES, CONFIGURATION
None.
.SH BUGS
Additional input devices other than the keyboard and mouse are not
disabled.

The timeouts, bitmaps and mouse cursor colour cannot be modifed.
.SH SEE ALSO
.BR X "(1), " xlock "(1), Xlib Documentation."
.SH AUTHOR
Mathias Gumz <akira at gmail dot com>
Ian Jackson <ijackson@nyx.cs.du.edu>
.SH COPYRIGHT
Copyright (C) 2005 Mathias Gumz
Copyright (C) 1993, 1994 by Ian Jackson.

Permission to use, copy, modify and distribute this software and its
documentation is granted under the terms of the GNU General Public
Licence, version 2 or later, as published by the Free Software
Foundation.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.


CHANGELOG
Version 0.2

2005-05-06:

	* added -nolock
	* discard all pending xevents after wrong password to prevent
	  strange folks for blocking the current session by filling
	  the eventqueue with wrong authentification and thus block
	  the real authentification (kind of denial of service :)
	* added suuport to load / have different cursor-shapes
		+ -cursor theme|<format>:<filename>
	* added support for xcursor
	* restructuring
		+ contrib/ -> for stuff coming from the "outside"
		  contrib/xcursor-watch -> xcursor taken from 5532-BlueGlass-XCursors-3D-0.4.tar.bz2
		+ src/

2005-05-02:

	* restructuring

2005-04-30:

	* added -cursor <theme>
	* cursor-bitmap-themes are stored in bitmaps/
	* started svn-usage

Version 0.1

2005-04-28: 
	* first release (0.1)
	* added pam-support (works ok with linux + freebsd so far)
	* use scons for building the stuff

# vim:ft=changelog


src/aklock.c
/*------------------------------------------------------------------*\

    file: aklock.c

        X Transparent Lock

    copyright:

        Copyright (C)1993,1994 Ian Jackson   (xtrlock)
        Copyright (C)2005 Mathias Gumz (forked aklock)

    license:

        This is free software; you can redistribute it and/or modify
        it under the terms of the GNU General Public License as published by
        the Free Software Foundation; either version 2, or (at your option)
        any later version.

        This is distributed in the hope that it will be useful,
        but WITHOUT ANY WARRANTY; without even the implied warranty of
        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
        GNU General Public License for more details.

\*------------------------------------------------------------------*/

/*------------------------------------------------------------------*\
\*------------------------------------------------------------------*/
#include <X11/X.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/keysym.h>
#include <X11/Xos.h>

#ifdef HAVE_XPM
#   include <X11/xpm.h>
#endif /* HAVE_XPM */
#ifdef HAVE_XCURSOR
#   include <X11/Xcursor/Xcursor.h>
#endif /* HAVE_XCURSOR */

#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <errno.h>
#include <pwd.h>
#include <grp.h>
#include <limits.h>
#include <string.h>
#include <unistd.h>
#include <math.h>
#include <ctype.h>

#ifdef SHADOW_PWD
#    include <shadow.h>
#endif /* SHADOW_PWD */

#ifdef PAM_PWD
#    include <security/pam_appl.h>
#    ifdef LINUX
#        include <security/pam_misc.h>
#    endif
#endif /* PAM_PWD */

/*----------------------------------------------*\
\*----------------------------------------------*/

#include "aklock.h"

/*------------------------------------------------------------------*\
    globals
\*------------------------------------------------------------------*/

#define TIMEOUTPERATTEMPT 30000
#define MAXGOODWILL  (TIMEOUTPERATTEMPT*5)
#define INITIALGOODWILL MAXGOODWILL
#define GOODWILLPORTION 0.3

#ifdef DEBUG
#    define DBGMSG fprintf(stderr, "%s : %d\n", __FUNCTION__, __LINE__); fflush(stderr)
#else
#    define DBGMSG
#endif // DEBUG

struct passwd *pw;



/*------------------------------------------------------------------*\
\*------------------------------------------------------------------*/

void displayUsage() {
    printf("\naklock [-h] [-v] [-blank] [-nolock] [-cursor <theme|xcursor:file>]\n");
}





/*------------------------------------------------------------------*\
    pam-related stuff

    taken from pure-ftpd's authstuff, but you can see similar stuff
    in xlockmore, openssh and basicly all pam-related apps :)
\*------------------------------------------------------------------*/
#ifdef PAM_PWD
#    define PAM_YN { \
                if (PAM_error != 0 || pam_error != PAM_SUCCESS) { \
                    fprintf(stderr, "pam error:%s\n", pam_strerror(pam_handle, pam_error)); \
                    pam_end(pam_handle, pam_error); \
                    PAM_username = NULL; \
                    PAM_password = NULL; \
                return 0;\
            } \
     }

#   define GET_MEM \
       size += sizeof(struct pam_response); \
       if ((reply = realloc(reply, size)) == NULL) { \
           PAM_error = 1; \
           return PAM_CONV_ERR; \
       }

static const char* PAM_username = NULL;
static const char* PAM_password = NULL;
static int PAM_error = 0;
static int pam_error = PAM_SUCCESS;

static int PAM_conv(int num_msg, const struct pam_message **msgs,
	              struct pam_response **resp, void *appdata_ptr) {

    int count = 0;
    unsigned int replies = 0U;
    struct pam_response *reply = NULL;
    size_t size = (size_t) 0U;

    (void) appdata_ptr;
    *resp = NULL;
    for (count = 0; count < num_msg; count++) {
        switch (msgs[count]->msg_style) {
        case PAM_PROMPT_ECHO_ON:
            GET_MEM;
            memset(&reply[replies], 0, sizeof reply[replies]);
            if ((reply[replies].resp = strdup(PAM_username)) == NULL) {
#    ifdef PAM_BUF_ERR
                reply[replies].resp_retcode = PAM_BUF_ERR;
#    endif
                PAM_error = 1;
                return PAM_CONV_ERR;
            }
            reply[replies++].resp_retcode = PAM_SUCCESS;
            /* PAM frees resp */
            break;
        case PAM_PROMPT_ECHO_OFF:
            GET_MEM;
            memset(&reply[replies], 0, sizeof reply[replies]);
            if ((reply[replies].resp = strdup(PAM_password)) == NULL) {
#    ifdef PAM_BUF_ERR
                reply[replies].resp_retcode = PAM_BUF_ERR;
#    endif
                PAM_error = 1;
                return PAM_CONV_ERR;
            }
            reply[replies++].resp_retcode = PAM_SUCCESS;
            /* PAM frees resp */
            break;
        case PAM_TEXT_INFO:
            /* ignore it... */
            break;
        case PAM_ERROR_MSG:
        default:
            /* Must be an error of some sort... */
            free(reply);
            PAM_error = 1;
            return PAM_CONV_ERR;
        }
    }
    *resp = reply;
    return PAM_SUCCESS;
}

static struct pam_conv PAM_conversation = {
    &PAM_conv, NULL
};
#endif /* PAM_PWD */
/*------------------------------------------------------------------*\
\*------------------------------------------------------------------*/

int passwordOk(const char *s) {
#if 0
    char key[3];
    char *encr;

    key[0] = *(pw->pw_passwd);
    key[1] = (pw->pw_passwd)[1];
    key[2] = 0;
    encr = crypt(s, key);
    return !strcmp(encr, pw->pw_passwd);
#else
    /* simpler, and should work with crypt() algorithms using longer
       salt strings (like the md5-based one on freebsd).  --marekm */
#ifdef PAM_PWD
    pam_handle_t* pam_handle = NULL;
    PAM_username = pw->pw_name;
    PAM_password = s;
    pam_error = pam_start("login", PAM_username, &PAM_conversation, &pam_handle);
    PAM_YN;
    pam_error = pam_authenticate(pam_handle, 0);
    PAM_YN;
    pam_error = pam_end(pam_handle, pam_error);
    PAM_YN;
    return 1;
#else
    return !strcmp(crypt(s, pw->pw_passwd), pw->pw_passwd);
#endif /* PAM_PWD */
#endif /* 0 */
}

/*------------------------------------------------------------------*\
    check if the system would be able to authentificate the user
\*------------------------------------------------------------------*/
void checkAuth() {

#ifdef SHADOW_PWD
    struct spwd *sp;
#endif

    errno = 0;
    pw = getpwuid(getuid());
    if (!pw) {
        perror("password entry for uid not found");
        exit(1);
    }

#ifdef SHADOW_PWD
    sp = getspnam(pw->pw_name);
    if (sp)
        pw->pw_passwd = sp->sp_pwdp;
    endspent();
#endif
    /* we can be installed setuid root to support shadow passwords,
       and we don't need root privileges any longer.  --marekm */
    setuid(getuid());
#ifndef PAM_PWD
    if (strlen(pw->pw_passwd) < 13) {
        perror("aklock: password entry has no pwd\n");
        exit(1);
    }
#endif /* PAM_PWD */

}



/*------------------------------------------------------------------*\
\*------------------------------------------------------------------*/


/*------------------------------------------------------------------*\
\*------------------------------------------------------------------*/

void initOpts(struct akOpts* opts) {

    opts->dont_lock = 0;
    opts->use_blank = 0;

    opts->cursor_name = "mini";

    opts->color_bg = "steelblue3";
    opts->color_fg = "grey25";

}

void initXInfo(struct akXInfo* xinfo, struct akOpts* opts) {


    Display* dpy = XOpenDisplay(NULL);
    XColor color_fg;
    XColor color_bg;
    XColor tmp_color;
    Pixmap pixmap_cursor;
    Pixmap pixmap_cursor_mask;
    XWindowAttributes xgwa;
    struct akCursor* cursor = NULL;

    if (!dpy) {
        perror("aklock: error, can't open connection to X");
        exit(1);
    }

    xinfo->display = dpy;
    xinfo->window = 0;

    xinfo->root = DefaultRootWindow(dpy);
    xinfo->colormap = DefaultColormap(dpy, DefaultScreen(dpy));

    XGetWindowAttributes(dpy, xinfo->root, &xgwa);

    if (opts->use_blank) {
        xinfo->width = xgwa.width;
        xinfo->height = xgwa.height;
    } else {
        xinfo->width = 1;
        xinfo->height = 1;
    }

    if((XAllocNamedColor(dpy, xinfo->colormap, opts->color_bg, &tmp_color, &color_bg)) == 0)
        XAllocNamedColor(dpy, xinfo->colormap, "black", &tmp_color, &color_bg);
    if((XAllocNamedColor(dpy, xinfo->colormap, opts->color_fg, &tmp_color, &color_fg)) == 0)
        XAllocNamedColor(dpy, xinfo->colormap, "white", &tmp_color, &color_fg);

    /* load cursors */
#ifdef HAVE_XCURSOR
    if (opts->cursor_name && (strstr(opts->cursor_name, "xcursor:"))) {
        Cursor xcursor;
        if ((xcursor = XcursorFilenameLoadCursor(dpy, &opts->cursor_name[8]))) {
            xinfo->cursor = xcursor;
            return;
        } else {
            printf("aklock: error, couldnt load [%s]\n", &opts->cursor_name[8]);
        }
    }
#endif /* HAVE_XCURSOR */

    /* TODO: doesnt work yet. */
/*
 * if (opts->cursor_name && (strstr(opts->cursor_name, "xbm:"))) {
        unsigned int w, h, xhot, yhot;
        if (XReadBitmapFile(dpy, xinfo->root, &opts->cursor_name[4],
                            &w, &h, &pixmap_cursor, &xhot, &yhot)) {
            xinfo->cursor = XCreatePixmapCursor(dpy,
                                                pixmap_cursor, NULL,
                                                &color_fg, &color_bg,
                                                xhot, yhot);
            return;
        } else {
            printf("aklock: error, couldnt load [%s]\n", &opts->cursor_name[4]);
        }
    }
   */
    /* look internal cursors */
    for(cursor = ak_cursors; cursor->name != NULL; cursor++) {
        if (!strcmp(cursor->name, opts->cursor_name))
                break;
    }

    if (!cursor->name)
        cursor = ak_cursors;

    pixmap_cursor = XCreateBitmapFromData(dpy, xinfo->root, cursor->bits, cursor->width, cursor->height);
    pixmap_cursor_mask = XCreateBitmapFromData(dpy, xinfo->root, cursor->mask, cursor->width, cursor->height);

    xinfo->cursor = XCreatePixmapCursor(dpy,
                                        pixmap_cursor, pixmap_cursor_mask,
                                        &color_fg, &color_bg,
                                        cursor->x_hot, cursor->y_hot);


}

int main(int argc, char **argv) {

    XEvent ev;
    KeySym ks;
    char cbuf[10], rbuf[50];
    int clen, rlen = 0;
    long goodwill = INITIALGOODWILL, timeout = 0;

    XSetWindowAttributes xswa;
    long xsmask = 0;

    struct akXInfo xinfo;
    struct akOpts opts;

    int arg = 0;

    initOpts(&opts);

    // parse options
    if (argc != 1) {
        for(arg = 1; arg <= argc; arg++) {
            if (!strcmp(argv[arg - 1], "-blank")) {
                opts.use_blank = 1;
            } else if (!strcmp(argv[arg - 1], "-nolock")) {
                opts.dont_lock = 1;
            } else if (!strcmp(argv[arg - 1], "-cursor")) {
                if (arg < argc)
                    opts.cursor_name = argv[arg];
                else {
                    printf("aklock: error, missing argument\n");
                    displayUsage();
                    exit(1);
                }
            } else if (!strcmp(argv[arg - 1], "-h")) {
                displayUsage();
                exit(0);
            } else if (!strcmp(argv[arg - 1], "-v")) {
                printf("aklock-%s by m.gumz 2005\n", VERSION);
                exit(0);
            }
        }
    }

    if (!opts.dont_lock)
        checkAuth();

    initXInfo(&xinfo, &opts);


    /* create the windows */
    xswa.override_redirect = True;
    xsmask |= CWOverrideRedirect;

    if (opts.use_blank) {

        xswa.background_pixel = BlackPixel(xinfo.display, DefaultScreen(xinfo.display));
        xswa.colormap = xinfo.colormap;
        xsmask |= CWBackPixel;
        xsmask |= CWColormap;
    }

    xinfo.window = XCreateWindow(xinfo.display, xinfo.root,
                          0, 0, xinfo.width, xinfo.height,
                          0, /* borderwidth */
                          CopyFromParent, /* depth */
                          InputOutput, /* class */
                          CopyFromParent, /* visual */
                          xsmask, &xswa);



    XSelectInput(xinfo.display, xinfo.window, KeyPressMask|KeyReleaseMask);
    XMapWindow(xinfo.display, xinfo.window);
    XRaiseWindow(xinfo.display, xinfo.window);
    
    /* TODO: -bg <blank|transparent|shaded> */
    if (opts.use_blank) {
        XImage* ximage;
                
        ximage = XGetImage (xinfo.display, xinfo.root, 0, 0,
                    xinfo.width, xinfo.height, AllPlanes, ZPixmap);

        if (ximage) {
            GC gc;
            XGCValues xgcv;
            xgcv.background = BlackPixel(xinfo.display, DefaultScreen(xinfo.display));
            xgcv.foreground = 0;
            
            gc = XCreateGC(xinfo.display, xinfo.window, GCForeground|GCBackground, &xgcv);

            XPutImage(xinfo.display, xinfo.window, 
                      gc, 
                      ximage, 
                      0, 0, 
                      0, 0, xinfo.width, xinfo.height);
            XDestroyImage(ximage);
            XFreeGC(xinfo.display, gc);
        }
    }
    /* try to grab 2 times, another process (windowmanager) may have grabbed
     * the keyboard already */
    if ((XGrabKeyboard(xinfo.display, xinfo.window, True, GrabModeAsync, GrabModeAsync,
                      CurrentTime)) != GrabSuccess) {
        sleep(1);
        if ((XGrabKeyboard(xinfo.display, xinfo.window, True, GrabModeAsync, GrabModeAsync,
                        CurrentTime)) != GrabSuccess) {
            perror("aklock: couldnt grab the keyboard.\n");
            exit(1);
        }
    }

    if (XGrabPointer(xinfo.display, xinfo.window, False, (KeyPressMask|KeyReleaseMask) & 0,
                     GrabModeAsync, GrabModeAsync, None, xinfo.cursor, CurrentTime) != GrabSuccess) {
        XUngrabKeyboard(xinfo.display, CurrentTime);
        perror("aklock: couldnt grab the pointer.\n");
        exit(1);
    }

    /* eventhandling */
    for(;;) {
        XNextEvent(xinfo.display, &ev);
        switch (ev.type) {
        case KeyPress:
            
            if (opts.dont_lock)
                goto exit;

            if (ev.xkey.time < timeout) {
                XBell(xinfo.display, 0);
                break;
            }

            clen = XLookupString(&ev.xkey, cbuf, 9, &ks, 0);
            switch (ks) {
            case XK_Escape:
            case XK_Clear:
                rlen = 0;
                break;
            case XK_Delete:
            case XK_BackSpace:
                if (rlen > 0)
                    rlen--;
                break;
            case XK_Linefeed:
            case XK_Return:
                if (rlen == 0)
                    break;
                rbuf[rlen] = 0;

                if (passwordOk(rbuf))
                    goto exit;
                
                XSync(xinfo.display, True); /* discard pending events to start really fresh */
                XBell(xinfo.display, 0);
                rlen = 0;
                if (timeout) {
                    goodwill += ev.xkey.time - timeout;
                    if (goodwill > MAXGOODWILL) {
                        goodwill = MAXGOODWILL;
                    }
                }
                timeout = -goodwill * GOODWILLPORTION;
                goodwill += timeout;
                timeout += ev.xkey.time + TIMEOUTPERATTEMPT;
                break;
            default:
                if (clen != 1)
                    break;
                if (rlen < sizeof(rbuf))
                    rbuf[rlen] = cbuf[0];
                rlen++;
                break;
            }
            break;
        default:
                break;
        }
    }

exit:
    XDestroyWindow(xinfo.display, xinfo.window);
    XFreeCursor(xinfo.display, xinfo.cursor);
    XCloseDisplay(xinfo.display);

    return 0;
}

